### DISCUSSION TITLE
Why not using sqlparser ast directly?

### DISCUSSION THREAD

**Discussion by barsdeveloper** [gluesql/gluesql#1588] [Q&A] [answered]
Hello, Is there any reason why you implemented a ast instead of using the one in sqlparser? It seems to me that they are very similar in the sense that they provide the similar information.
From what I see sqlparser's ast provides even more information, compare for example the create table statement:
https://docs.rs/gluesql/latest/gluesql/core/ast/enum.Statement.html#variant.CreateTable
https://docs.rs/sqlparser/latest/sqlparser/ast/struct.CreateTable.html

I also have an additional question: why ToSql trait does not support the sql dialects?
https://docs.rs/gluesql/latest/gluesql/core/ast/trait.ToSql.html
If I have a specific storage, then I should be able to generate a query specifically for that database

**Comment by panarch** [comment, marked as answer]
> From what I see sqlparser's AST provides even more information, compare for example the create table statement:

That’s a great observation, but having "even more information" is actually the issue. The `sqlparser-rs` library is designed to support the SQL syntax of various databases simultaneously. GlueSQL, however, is just one of those databases. If we were to directly use the AST from `sqlparser` in GlueSQL, we would have to handle numerous AST nodes that GlueSQL doesn’t currently support—or doesn’t plan to support—at runtime. This would add unnecessary complexity by requiring additional error handling.

To address this, GlueSQL uses its `translate` module to extract only the relevant parts of the `sqlparser` AST needed for its purposes.

Additionally, there’s another reason: GlueSQL offers the flexibility to optionally use a query planner. While most databases have a mandatory query planner, GlueSQL prioritizes supporting diverse storage backends and environments. Designing a universal query planner for all potential use cases isn’t feasible, so GlueSQL allows developers to either use or bypass the planner.

As a result, GlueSQL's AST needs to include execution-related information beyond what is typically found in a parser's AST. For instance, GlueSQL’s AST includes a default execution plan by design. When an optional query planner is applied, it takes the AST as input and returns an enhanced AST with an optimized execution plan. 

This separation makes `sqlparser` an excellent library for parsing but less suitable for execution purposes, leading to the need for GlueSQL’s custom AST, which incorporates execution details alongside the parsing results.

---

> I also have an additional question: why does the `ToSql` trait not support SQL dialects? If I have a specific storage, then I should be able to generate a query specifically for that database.

GlueSQL's primary goal is to provide a unified SQL interface for interacting with diverse storage backends. Therefore, it intentionally does not support dialects. The focus is not on preserving storage-specific syntax but rather on providing a consistent way to manage different storage types through a single SQL or query builder interface.

**Reply by barsdeveloper** [reply]
Thank you very much for taking time to write such a detailed answer.
